lib/open_api_spex/open_api/map.ex

```diff
@@ -0,0 +1,395 @@
+defmodule OpenApiSpex.OpenApi.Map do
+  @moduledoc """
+  This module exposes functionality to convert an arbitrary map into a OpenApi struct.
+  """
+  alias OpenApiSpex.{
+    Components,
+    Contact,
+    Discriminator,
+    Encoding,
+    Example,
+    ExternalDocumentation,
+    Header,
+    Info,
+    License,
+    Link,
+    MediaType,
+    OAuthFlow,
+    OAuthFlows,
+    OpenApi,
+    Operation,
+    Parameter,
+    PathItem,
+    Reference,
+    RequestBody,
+    Response,
+    Schema,
+    SecurityScheme,
+    Server,
+    ServerVariable,
+    Tag,
+    Xml
+  }
+
+  def decode(%{openapi: _openapi, info: _info, paths: _paths} = map) do
+    map
+    |> to_struct(OpenApi)
+    |> to_struct(:info, Info)
+    |> to_struct(:paths, PathItems)
+    |> to_struct(:servers, Servers)
+    |> to_struct(:components, Components)
+    |> to_struct(:tags, Tag)
+    |> to_struct(:externalDocs, ExternalDocumentation)
+  end
+
+  defp embedded_ref_or_struct(list, mod) when is_list(list) do
+    list
+    |> Enum.map(fn
+      %{"$ref": _} = v -> struct(Reference, v)
+      v -> to_struct(v, mod)
+    end)
+  end
+
+  defp embedded_ref_or_struct(map, mod) when is_map(map) do
+    map
+    |> Map.new(fn
+      {k, %{"$ref": _} = v} ->
+        {k, struct(Reference, v)}
+
+      {k, v} ->
+        {k, to_struct(v, mod)}
+    end)
+  end
+
+  defp convert_value_to_atom_if_present(map, key) do
+    if Map.has_key?(map, key) do
+      Map.update!(map, key, fn
+        nil ->
+          nil
+
+        v ->
+          String.to_atom(v)
+      end)
+    else
+      map
+    end
+  end
+
+  defp prepare_schema(map) do
+    map
+    |> convert_value_to_atom_if_present(:type)
+    |> to_list_of_atoms(:required)
+  end
+
+  defp to_list_of_atoms(map, key) do
+    if Map.has_key?(map, key) do
+      Map.update!(map, key, fn
+        nil ->
+          nil
+
+        v ->
+          v
+          |> Enum.map(&String.to_atom/1)
+      end)
+    else
+      map
+    end
+  end
+
+  defp to_struct(nil, _mod), do: nil
+
+  defp to_struct(tag, Tag) when is_binary(tag), do: tag
+
+  defp to_struct(map, Tag) when is_map(map) do
+    Tag
+    |> struct(map)
+    |> to_struct(:externalDocs, ExternalDocumentation)
+  end
+
+  defp to_struct(list, Tag) when is_list(list) do
+    list
+    |> Enum.map(&to_struct(&1, Tag))
+  end
+
+  defp to_struct(map, Components) do
+    # NOTE: need to conver all keys to strings at a lower level
+    map =
+      Map.new(map, fn
+        {k, v} -> {k, v |> Map.new(fn {k, v} -> {Atom.to_string(k), v} end)}
+      end)
+
+    Components
+    |> struct(map)
```

Comment:
I believe Elixir's built-in `struct` does not accept string keys. So this is potentially not doing anything besides `%Components{}`.